core/Test/Tasty/Parallel.hs

```diff
@@ -52,9 +53,9 @@ runInParallel nthreads actions = do
 
   actionsVar <- atomically $ newTMVar actions
 
-  pids <- replicateM nthreads (forkIO $ work actionsVar)
+  pids <- replicateM nthreads (async $ work actionsVar)
 
-  return $ mapM_ killThread pids
+  return $ mapM_ cancel pids
 
 work :: TMVar [Action] -> IO ()
 work actionsVar = go
```

Comment:
I tried replacing the `actionRun this :: IO ()` below with the contents of my test, and that fixed the problem, `defaultMain` terminates again. Which means that `actionRun this` is not simply running one of the tests, as I expected, it is instead doing something more complicated. I am still reading the code and #48 in order to figure out what it is that `actionRun this` is doing instead.

Reply:
I now understand that `actionRun this` does run one of the tests, but it runs it inside a `try`, and so `Async.cancel` does not cause the worker thread to terminate.